### PR TITLE
[clang-cpp-frontend] Added support for -CXXConstructExpr

### PR DESCRIPTION
**Pull request by kunjsong01**
This PR added support for `CXXConstructExpr`.

### Test case: 
```cpp
#include <assert.h>
class Test {
public:
  int x;
  /*
  Test(int _x) {
    x = _x; // requires CXXThisExpr
  }
  */
  int increment_x(int a) {
    return a + 1;
  }
};

int main() {
  Test testObj;
  int y = 0;
  y = testObj.increment_x(y);
  assert(y != 1);
  return 0;
}
```

### TODOs: 
1. The `CXXMemberCallExpr` does not generate the correct symbol: 
```
Symbol......: c:@F@main#
Module......: main
Base name...: main
Mode........: C (0)
Type........: signed int ()
Value.......:
{
Test testObj={ .x=0 };

signed int y=0;

y = member
  * type: code
      * arguments:
  * operands:
    0: symbol
        * type: struct
            * tag: Test
            * components:
              0: component
                  * type: signedbv
                      * width: 32
                      * #cpp_type: signed_int
                  * name: x
                  * pretty_name: x
        * name: testObj
        * identifier: c:main.cpp@185@F@main#@testObj
        * #location:
          * file: main.cpp
          * line: 18
          * function: main
        * #lvalue: 1
  * component_name:
  * #location:
    * file: main.cpp
    * line: 18
    * function: main(testObj, y);

y != 1 ? 0 : __assert_fail(&"y != 1"[0], &"main.cpp"[0], 19, &"int main()"[0]);

return 0;

}
Flags.......: lvalue
Location....: file main.cpp line 15
```

Expected the memebr function call to be like `y=increment_x(&testObj, y)`.

### CODE REVIEW

**Comment by mikhailramalho** [comment]
Nice, this has been missing from esbmc for quite some time!

I implemented a version that covers most of the cases some time ago, but there are a lot of corner cases that need handling.

btw, at goto conversion, constructors are expected to be transformed from:

`MyClass foo(a, b, c)`

to:

```
MyClass foo;
MyClass::constructor(&foo, a, b, c);
```

I'll search for the code I wrote and try to upload it here.

**Comment by monicamarcus** [comment]
Not knowing much about your project, I read the Test Case above. I wonder what is the meaning of x in the following function definition:

int increment_x(int a) {
    return a + 1;
  }

At first I thought the result of the function call should depend on x. But maybe x has some other role.

**Comment by kunjsong01** [comment]
> At first I thought the result of the function call should depend on x. But maybe x has some other role.

My apologies for the confusing function name. Initially, it was named "increment_x" to test method that modifies a class field. But I found that it also requires support for other AST nodes. It should be "increment_a" because the function only returns (a+1).

**Comment by kunjsong01** [comment]
### Investigations of the failed test cases:
```
The following tests FAILED:
	1689 - regression/esbmc-cpp/cbmc/Templates14 (Failed)
	1691 - regression/esbmc-cpp/cbmc/Templates16 (Failed)
	1692 - regression/esbmc-cpp/cbmc/Templates17 (Failed)
	1693 - regression/esbmc-cpp/cbmc/Templates18 (Failed)
	1695 - regression/esbmc-cpp/cbmc/Templates2 (Failed)
	1711 - regression/esbmc-cpp/cbmc/Templates34 (Failed)
	1713 - regression/esbmc-cpp/cbmc/Templates36 (Failed)
	1714 - regression/esbmc-cpp/cbmc/Templates37 (Failed)
	1716 - regression/esbmc-cpp/cbmc/Templates39 (Failed)
	1720 - regression/esbmc-cpp/cbmc/Templates7 (Failed)
```
The test cases above are false failures. They actually passed "VERIFICATION SUCCESSFUL" as a result of this PR that enables conversion of zero-argument implicit constructor. 
```
ESBMC version 6.9.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmc libc
file Templates2/main.cpp: Parsing
Converting
Generating GOTO Program
GOTO program creation time: 0.147s
GOTO program processing time: 0.000s
Starting Bounded Model Checking
Symex completed in: 0.002s (12 assignments)
Slicing time: 0.000s (removed 12 assignments)
Generated 1 VCC(s), 0 remaining after simplification (0 assignments)
BMC program time: 0.002s

VERIFICATION SUCCESSFUL
```
**Without this PR,** they would report conversion error due to unsupported clang expr `CXXConstructExpr`: 
```
ESBMC version 6.9.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmc libc
file Templates2/main.cpp: Parsing
Converting
### ERROR:
Conversion of unsupported clang expr: "CXXConstructExpr" to expression
CXXConstructExpr 0x8d48840 <Templates2/main.cpp:12:16> 'int_array<3>':'int_array<3>' 'void () throw()'

### ERROR:
CONVERSION ERROR
```

The reason they were reported "failed" in our CI is because they were labelled "KNOWNBUG" in the `test.desc`, which expected the test case to fail. 

TODOs:
1. ~~Replace "KNOWNBUG" with "THOROUGH" in those test cases~~ (Done)

**Comment by kunjsong01** [comment]
With this PR, we are now capable of doing a small demo that involves class, tempalte and array: 
```cpp
#include<cassert>

template <int size>
class int_array
{
public:
  int array[size];
};

int main()
{
  int_array<3> a;
  a.array[0]=1;
  a.array[1]=1;
  a.array[2]=1;

  int i = 0;
  while(i < 3)
  {
    a.array[i]=1;
    ++i;
  }

  for (int j = 0; j <= 3; ++j)
  {
    assert(a.array[j]==1);
  }
}
```
`./esbmc Templates2/main.cpp -I ~/libraries/`
```
State 12 file main.cpp line 24 function main thread 0
----------------------------------------------------
  j = 0 (00000000 00000000 00000000 00000000)

State 13 file main.cpp line 24 function main thread 0
----------------------------------------------------
  j = 1 (00000000 00000000 00000000 00000001)

State 14 file main.cpp line 24 function main thread 0
----------------------------------------------------
  j = 2 (00000000 00000000 00000000 00000010)

State 15 file main.cpp line 24 function main thread 0
----------------------------------------------------
  j = 3 (00000000 00000000 00000000 00000011)

State 16 file main.cpp line 26 function main thread 0
----------------------------------------------------
Violated property:
  file main.cpp line 26 function main
  array bounds violated: array upper bound
  (signed long int)j < 3


VERIFICATION FAILED
```

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 1]
```diff
@@ -540,6 +540,29 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
...
 
+  case clang::Stmt::CXXConstructExprClass:
+  {
+    // Reference to a declared construstor
+    const clang::CXXConstructExpr &ctr_call =
+      static_cast<const clang::CXXConstructExpr &>(stmt);
+
+    if(ctr_call.getNumArgs() != 0)
+    {
+      assert(
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:552]: Shall we mark it as TODO and add some comments?

**Review by lucasccordeiro — APPROVED** [review]
Overall, this PR looks good to me.

**Comment by kunjsong01** [comment]
### Some notes for next steps:
Although this PR just enables a demo example, it's a good starting point to extend the clang-cpp frontend in a step-by-step manner, towards eventually a feature-complete ESBMC++. 

As Mikhail mentioned in [the comment above](https://github.com/esbmc/esbmc/pull/686#issuecomment-1107947298), I still need to find a better way to convert constructors properly and teach our GOTO IR to avoid spaghetti code. 

### TODOs:
We are going to follow a TDD approach for future TODOs: 
1. Audit examples from C++ How to Program book and extract the examples from chapters that address basic language constructs of C++11.   [Deliverable: Add C++11 regression suite]
2. Pass the newly added C++11 regression suite 100%. [Deliverable: More PRs for clang-cpp-frontend]

**Comment by monicamarcus** [comment]
Hi @kunjsong01 @lucasccordeiro 

I am glad there is so much progress in the work for esbmc++

What I would like to verify is a C++ program that allows for using intrusive double linked lists. All the C++ code is written and verified using unit tests written manually by Liviu Ionescu https://github.com/ilg-ul

I started to modify the code to use C++03 instead of C++20. I hope I can tweak it such that the functionality is not lost.

My goal is to prove that esbmc++ is much better to verify C++ code than writing unit tests manually.

Best wishes,
Monica

**Comment by lucasccordeiro** [comment]
Hi @monicamarcus,

That's great! You could write a blog in our GitHub repository about this work.

Can I ask you whether you could share this implementation of the double liked list in C++ with us?

Many thanks,
Lucas

**Comment by monicamarcus** [comment]
Hi @lucasccordeiro,

Yes, I shall write a blog as you suggested and will probably do most of  this work over weekend(s).

Here is the link to the C++ code: https://github.com/micro-os-plus/utils-lists-xpack/

Best wishes,
Monica

**Comment by monicamarcus** [comment]
Hi @lucasccordeiro ,

I started my [blog](https://github.com/users/monicamarcus/projects/1/views/1) there is not much to see there yet but I wanted to know whether you could see the ReadMe content shown in the attached screenshot.

Screen Shot 2022-05-02 at 9 35 44 AM

Yesterday I did some experiments with the code for Liviu's lists regarding the constexpr keyword which is not in C++03. I stopped short of adding a few words in the blog but I shall add them.

Can I use now esbmc++ with the latest additions?

Best,
Monica

**Comment by kunjsong01** [comment]
Hi @monicamarcus 

Thanks a lot for writing a blog about ESBMC++!

The [source code](https://github.com/micro-os-plus/utils-lists-xpack/tree/xpack/src) appears to use APIs from other mOS++ headers files, such as [micro-os-plus/utils/lists.h](https://micro-os-plus.github.io/reference/cmsis-plus/lists_8h_source.html) and [micro-os-plus/diag/trace.h](https://micro-os-plus.github.io/reference/cmsis-plus/trace_8h_source.html). 

> Can I use now esbmc++ with the latest additions?

Unfortunately, I'm afraid clang-based ESBMC++ is still not feature complete. The aforementioned source code contains some language constructs that are not supported by our clang-based C++ frontend, e.g. the move assignment operator.  
Hopefully, by Q4 this year we will have a clang-based ESBMC++ to verify your program. 

Many thanks,
Kunjian

**Comment by monicamarcus** [comment]
Hi @kunjsong01,

Thank you for your message. My blog is not going to be about esbmc++. What I intend is to work on the source code for the lists and tweak it such that it can be understood by C++03 or C++11 compilers without loosing in lists functionality. My blog is going to be about this work, whether successful or not.

So can you please let me know whether the latest additions to esbmc/esbmc++ make it feature complete for either C++03 or C++11? And what C++ version do you target for esbmc by the end of this year?

Thanks and best wishes,
Monica

**Comment by kunjsong01** [comment]
Hi @monicamarcus,

> So can you please let me know whether the latest additions to esbmc/esbmc++ make it feature complete for either C++03 or C++11?

I'm afraid it's not feature complete for C++11 yet. The special release mentioned in [previous comment](https://github.com/esbmc/esbmc/discussions/676#discussioncomment-2565367) is C++03 feature complete. 

> And what C++ version do you target for esbmc by the end of this year?

We target for C++11. 

Cheers,
Kunjian

**Comment by monicamarcus** [comment]
Hi @kunjsong01,

Thank you for your reply. 

I tried to install and build the esbmc special release which is feature complete for C++03 following the [instructions](https://github.com/esbmc/esbmc/blob/master/BUILDING.md) but I got some errors.  

I was not able to complete setting up Yices due to some error.

For  Z3 I needed to replace 'usr' by 'opt' in the path, because that is where I got Z3 installed using brew.

Also, I was not able to complete building esbmc due to some other error. 

I work on a MacBook computer using the Apple chip. I shall try to solve those errors by myself, but if you could take a look please let me know.

Thanks and best wishes,
Monica

**Comment by monicamarcus** [comment]
There are too many errors when trying to build esbmc, following the [instructions](https://github.com/esbmc/esbmc/blob/master/BUILDING.md) shown on your Github site. I give up trying to install it on my computer. I shall go back to just working to downgrade the intrusive lists code to C++03.

**Comment by kunjsong01** [comment]
Hi @monicamarcus,

> on a MacBook computer using the Apple chip.

AFAIK ESBMC does not work on aarch64 chip. 
The special release in the zipper folder mentioned in previous comment works on ubuntu 20.04 x86. 

Cheers,
Kunjian

**Comment by monicamarcus** [comment]
It is not the macbook computer that makes problems. I was able to install the dependencies. There are other issues with the paths and folders specified in your esbmc/CMakeLists.txt that are causing problems.

**Comment by lucasccordeiro** [comment]
Hello @monicamarcus,

If you intend to verify C++03 programs, you should use this version: https://ssvlab.github.io/esbmc/benchmarks/stvr_2021_experiments.zip

The old C++ frontend is not maintained for quite some time; I would say since 2015. So, even if you manage to compile ESBMC, you won't be able to verify C++ 03 programs with the current master.

Best,
Lucas

**Comment by monicamarcus** [comment]
Hi @lucasccordeiro 

Thank you for  the new link and information. I shall see what I can do.

Thanks and best wishes,
Monica

**Pull request merged**